### ISSUE TITLE
[JDBC persistence MariaDB] binding causes errors at "restoreOnStartup" restore for Location and Image Items

### ISSUE THREAD

**Issue by ohb1138** [openhab/openhab-addons#10494]
Hi there,
after upgrading from OH v2.5.12 directly to OH v3.1.0-M3 the JDBC persistence MariaDB binding causes errors in karaf.log when restoring Image and Location Items at startup. The following errors only occur once per item during startup phase. I have two examples for each:

Location Items:
```
2021-04-08 13:21:28.442 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: item is locationJR
2021-04-08 13:21:28.443 [DEBUG] [persistence.jdbc.internal.JdbcMapper] - JDBC::getHistItemFilterQuery filter='true' numberDecimalcount='3' table='Item372' item='locationJR (Type=LocationItem, State=NULL, Label=JR Location, Category=null)' itemName='locationJR'
2021-04-08 13:21:28.444 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::getHistItemFilterQueryProvider filter = org.openhab.core.persistence.FilterCriteria@17ac9388, numberDecimalcount = 3, table = Item372, simpleName = locationJR
2021-04-08 13:21:28.444 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::query queryString = SELECT time, value FROM Item372 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.445 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::doGetHistItemFilterQuery sql=SELECT time, value FROM Item372 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.448 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::ItemResultHandler::handleResult getState value = '49.9725155,8.3393236', unit = 'null', getClass = 'class java.lang.String', clazz = 'String'
2021-04-08 13:21:28.449 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: query for locationJR returned 1 rows in 6 ms
2021-04-08 13:21:28.450 [ERROR] [org.openhab.core.items.GenericItem  ] - Tried to set invalid state 51.5085,-0.1257 (StringType) on item locationJR of type LocationItem, ignoring it
…
2021-04-08 13:21:28.491 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: item is localStationLocation
2021-04-08 13:21:28.492 [DEBUG] [persistence.jdbc.internal.JdbcMapper] - JDBC::getHistItemFilterQuery filter='true' numberDecimalcount='3' table='Item40' item='localStationLocation (Type=LocationItem, State=NULL, Label=Location, Category=location)' itemName='localStationLocation'
2021-04-08 13:21:28.493 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::getHistItemFilterQueryProvider filter = org.openhab.core.persistence.FilterCriteria@7ac8c3b8, numberDecimalcount = 3, table = Item40, simpleName = localStationLocation
2021-04-08 13:21:28.494 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::query queryString = SELECT time, value FROM Item40 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.495 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::doGetHistItemFilterQuery sql=SELECT time, value FROM Item40 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.497 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::ItemResultHandler::handleResult getState value = '49.9725,8.3393', unit = 'null', getClass = 'class java.lang.String', clazz = 'String'
2021-04-08 13:21:28.498 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: query for localStationLocation returned 1 rows in 6 ms
2021-04-08 13:21:28.499 [ERROR] [org.openhab.core.items.GenericItem  ] - Tried to set invalid state 51.5085,-0.1257 (StringType) on item localStationLocation of type LocationItem, ignoring it
```

Image Items:

```
2021-04-08 13:21:28.292 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: item is MyCityCurrentConditionIcon
2021-04-08 13:21:28.292 [DEBUG] [persistence.jdbc.internal.JdbcMapper] - JDBC::getHistItemFilterQuery filter='true' numberDecimalcount='3' table='Item25' item='MyCityCurrentConditionIcon (Type=ImageItem, State=NULL, Label=Icon, Category=null)' itemName='MyCityCurrentConditionIcon'
2021-04-08 13:21:28.293 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::getHistItemFilterQueryProvider filter = org.openhab.core.persistence.FilterCriteria@1b63a0fa, numberDecimalcount = 3, table = Item25, simpleName = MyCityCurrentConditionIcon
2021-04-08 13:21:28.294 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::query queryString = SELECT time, value FROM Item25 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.295 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::doGetHistItemFilterQuery sql=SELECT time, value FROM Item25 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.298 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::ItemResultHandler::handleResult getState value = '[114, 97, 119, 32, 116, 121, 112, 101, 32, 40, 105, 109, 97, 103, 101, 47, 112, 110, 103, 41, 58, 32, 50, 55, 55, 51, 32, 98, 121, 116, 101, 115]', unit = 'null', getClass = 'class [B', clazz = 'byte[]'
2021-04-08 13:21:28.299 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'QueryablePersistenceService.query()' on 'org.openhab.persistence.jdbc.internal.JdbcPersistenceService@6bfdcace': class [B cannot be cast to class java.lang.String ([B and java.lang.String are in module java.base of loader 'bootstrap')
2021-04-08 13:21:28.317 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::getName: returning name 'jdbc' for queryable persistence service.
2021-04-08 13:21:28.317 [ERROR] [ence.internal.PersistenceManagerImpl] - Exception occurred while querying persistence service 'jdbc': class [B cannot be cast to class java.lang.String ([B and java.lang.String are in module java.base of loader 'bootstrap')
…
2021-04-08 13:21:28.583 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: item is MyCityHourlyForecast03ConditionIcon
2021-04-08 13:21:28.584 [DEBUG] [persistence.jdbc.internal.JdbcMapper] - JDBC::getHistItemFilterQuery filter='true' numberDecimalcount='3' table='Item45' item='MyCityHourlyForecast03ConditionIcon (Type=ImageItem, State=NULL, Label=Icon, Category=null)' itemName='MyCityHourlyForecast03ConditionIcon'
2021-04-08 13:21:28.584 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::getHistItemFilterQueryProvider filter = org.openhab.core.persistence.FilterCriteria@515081eb, numberDecimalcount = 3, table = Item45, simpleName = MyCityHourlyForecast03ConditionIcon
2021-04-08 13:21:28.585 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::query queryString = SELECT time, value FROM Item45 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.586 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::doGetHistItemFilterQuery sql=SELECT time, value FROM Item45 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.588 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::ItemResultHandler::handleResult getState value = '[114, 97, 119, 32, 116, 121, 112, 101, 32, 40, 105, 109, 97, 103, 101, 47, 112, 110, 103, 41, 58, 32, 50, 55, 55, 51, 32, 98, 121, 116, 101, 115]', unit = 'null', getClass = 'class [B', clazz = 'byte[]'
2021-04-08 13:21:28.589 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'QueryablePersistenceService.query()' on 'org.openhab.persistence.jdbc.internal.JdbcPersistenceService@6bfdcace': class [B cannot be cast to class java.lang.String ([B and java.lang.String are in module java.base of loader 'bootstrap')
2021-04-08 13:21:28.594 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::getName: returning name 'jdbc' for queryable persistence service.
2021-04-08 13:21:28.595 [ERROR] [ence.internal.PersistenceManagerImpl] - Exception occurred while querying persistence service 'jdbc': class [B cannot be cast to class java.lang.String ([B and java.lang.String are in module java.base of loader 'bootstrap')
```

I circumvented the issue by removing the "restoreOnStartup" option in the jdbc.persist and activated MapDB for this purpose.
I saw that this problem has already appeared in the past: https://github.com/openhab/openhab1-addons/issues/5761 and maybe it has slipped back into OH with the upgrade to V3.

I have investigated this issue in the openHAB community under the topic: [https://community.openhab.org/t/oh-3-1-m3-persistence-error-for-location-items-is-back-jdbc-restoreonstartup-throw-error-for-location-items/120430](url) and have been advised to open an issue here.

Thank you for considering a solution!
Cheers ohb1138

**Comment by ohb1138** [comment]
Sorry for pasting the error log messages into the post ... it is my first issue that I have raised ;-)

**Comment by cweitkamp** [comment]
Hi @ohb1138,

I submitted a fix for it. Are you able to test it?

See #10501

**Comment by ohb1138** [comment]
Hi @cweitkamp,
yes, sure ... probably not today but short term during the next days. Please provide a testing procedure.
Thank you for this awesome quick response!!!

**Comment by cweitkamp** [comment]
- Uninstall JDBC persistence service
- Download [mariadb-java-client-1.4.6.jar](https://repo1.maven.org/maven2/org/mariadb/jdbc/mariadb-java-client/1.4.6/mariadb-java-client-1.4.6.jar) and put it into your `addons/` folder
- Download new add-on version ([org.openhab.persistence.jdbc-3.1.0-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/6291569/org.openhab.persistence.jdbc-3.1.0-SNAPSHOT.zip)), extract the zip-file and put the *.jar file into your `addons/` folder

**Comment by ohb1138** [comment]
Hi @cweitkamp,

thank you for providing the fixes so quickly and please apologize the delayed response. I have just tested the fix applying the two persistence addons and the errors do not occur anymore but I receive the following warning:

`
09:35:10.529 [WARN ] [org.apache.felix.fileinstall         ] - Error while starting bundle: file:/openhab/addons/org.openhab.persistence.jdbc-3.1.0-SNAPSHOT.jar
org.osgi.framework.BundleException: Could not resolve module: org.openhab.persistence.jdbc [284]
  Unresolved requirement: Import-Package: javassist; version="[3.25.0,4.0.0)"; resolution:="optional"
  Unresolved requirement: Import-Package: javassist.bytecode; version="[3.25.0,4.0.0)"; resolution:="optional"
  Unresolved requirement: Import-Package: org.osgi.framework; version="[1.9.0,2.0.0)"

	at org.eclipse.osgi.container.Module.start(Module.java:444) ~[org.eclipse.osgi-3.12.100.jar:?]
	at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:383) ~[org.eclipse.osgi-3.12.100.jar:?]
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1260) [bundleFile:3.6.4]
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1233) [bundleFile:3.6.4]
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startAllBundles(DirectoryWatcher.java:1221) [bundleFile:3.6.4]
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:515) [bundleFile:3.6.4]
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:365) [bundleFile:3.6.4]
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:316) [bundleFile:3.6.4]
`
When this warning appeared the 

`
openhab> list -s | grep jdbc
284 │ Installed │  80 │ 3.1.0.202104101942      │ org.openhab.persistence.jdbc
286 │ Resolved    │  80 │ 1.4.6.0                 │ org.mariadb.jdbc
openhab>   
`

After a little while the mariadb persistence service switched to "Active":

`
openhab> list -s | grep jdbc
284 │ Installed │  80 │ 3.1.0.202104101942      │ org.openhab.persistence.jdbc
286 │ Active    │  80 │ 1.4.6.0                 │ org.mariadb.jdbc
openhab>   
`

Please advise if this needs further attention or if this warning can be neglected.

Cheers
Justus

**Comment by cweitkamp** [comment]
The warning is related to an OSGi Version updat in recent openHAB snapshot. This affects all bundles compiled from latest code. You will not be able to use them in older openHAB versions. To get them running you have to upgrade to latest openHAB nightly builds (e.g. build #2306 or higher).
